source/common/network/win32_socket_handle_impl.cc

```diff
@@ -84,5 +121,71 @@ void Win32SocketHandleImpl::reEnableEventBasedOnIOResult(const Api::IoCallUint64
   }
 }
 
+Api::IoCallUint64Result Win32SocketHandleImpl::drainToPeekBuffer() {
+  while (true) {
+    Buffer::Reservation reservation = peek_buffer_->reserveForRead();
+    Api::IoCallUint64Result result = IoSocketHandleImpl::readv(
+        reservation.length(), reservation.slices(), reservation.numSlices());
+    uint64_t bytes_to_commit = result.ok() ? result.return_value_ : 0;
+    reservation.commit(bytes_to_commit);
+    if (!result.ok() || bytes_to_commit == 0) {
+      return result;
+    }
+  }
+}
+
+Api::IoCallUint64Result Win32SocketHandleImpl::readFromPeekBuffer(void* buffer, size_t length) {
+  uint64_t copy_size = std::min(peek_buffer_->length(), static_cast<uint64_t>(length));
+  peek_buffer_->copyOut(0, copy_size, buffer);
+  peek_buffer_->drain(copy_size);
+  return Api::IoCallUint64Result(copy_size, Api::IoErrorPtr(nullptr, [](Api::IoError*) {}));
+}
+
+Api::IoCallUint64Result Win32SocketHandleImpl::readvFromPeekBuffer(uint64_t max_length,
+                                                                   Buffer::RawSlice* slices,
+                                                                   uint64_t num_slice) {
+  uint64_t total_length_to_read = std::min(max_length, peek_buffer_->length());
+  uint64_t num_slices_to_read = 0;
+  uint64_t num_bytes_to_read = 0;
+  for (; num_slices_to_read < num_slice && num_bytes_to_read < total_length_to_read;
+       num_slices_to_read++) {
+    auto length_to_copy = std::min(static_cast<uint64_t>(slices[num_slices_to_read].len_),
+                                   total_length_to_read - num_bytes_to_read);
+    peek_buffer_->copyOut(num_bytes_to_read, length_to_copy, slices[num_slices_to_read].mem_);
```

Comment:
It would be good to have a Buffer method that copies out into a range of slices.  Having this method in the buffer interface would centralize the testing of such code instead of resulting in ad-hoc implementations of the multi-slice copy out in multiple places in the envoy code.